### DISCUSSION TITLE
Support for <code> Tag Implemented but Missing UI Tool in WYSIWYG Field

### DISCUSSION THREAD

**Discussion by Abdallah-Awwad** [directus/directus#24693] [Feature Requests]
## Description:

While pasting some text, I discovered that the Directus WYSIWYG editor supports the `<code>` tag. However, there is no interface tool to apply this formatting directly from the UI.
![image](https://github.com/user-attachments/assets/396795f0-63fb-465a-97e7-da4024951b58)

Currently, users have to manually edit the HTML to use the `<code>` tag, as there is no icon or button in the WYSIWYG toolbar for this purpose.
![image](https://github.com/user-attachments/assets/cc3307df-b7c0-4f44-99ed-54740849dc45)

## Expected Behavior:
There should be an icon or button in the WYSIWYG toolbar that allows users to easily apply the `<code>` tag to selected text.

**Comment by rijkvanzanten** [comment]
TinyMCE doesn't have support for an inline-code toolbar button, so that would have to be implemented as a custom bit of functionality on our end.

**Comment by rijkvanzanten** [comment]
@Abdallah-Awwad Want to open a PR to implement this? 🙂

**Comment by Abdallah-Awwad** [comment]
Hey @rijkvanzanten, thanks for your response.
> @Abdallah-Awwad Want to open a PR to implement this? 🙂

Sure, I will give it a try!

**Comment by Natetronn** [comment]
Hi @rijkvanzanten 

@Abdallah-Awwad and I have worked on this PR together. We believe we have a working solution. 

Before we submit a PR, however, we do have a few questions and some things that should be hashed out before submission (unless you want us to hash those out after the fact, during the PR process?) For now, I'll explain what we've got:

In regards to translations, do we add the translation string/value to the en-US.yaml only for the Interface? Is that all that is required? Or do we add it to all the translations? If adding it to all, do we just leave its value in English, to later be translated on Crowdin or do we leave the value blank? What's the best practice for submitting code updates to Directus proper as it relates to translations? The docs didn't mention this under [Contributions](https://docs.directus.io/contributing/introduction.html) or [Directus Local Setup](https://docs.directus.io/contributing/running-locally.html).

We also have some decisions to make, as it relates to the icons used for both _Edit Source Code_ and our new _Inline Code_ toolbar icons.

Note: TinyCME only has these icons available (note the icon named `sourcecode`):
https://www.tiny.cloud/docs/tinymce/latest/editor-icon-identifiers/

TinyMCE use the same icon for both _Edit Source Code_ and _Inline Code_, but we don't have to, since we have a working icon replacement, via code, should we require one.

The soucecode `< >` icon is what is currently being used for _Edit Source Code_. That same icon, in our opinion, happens to also be suitable for Inline Code. The questions then arise:

- Do we use the same icon for both toolbar buttons? Our opinion was no, since that would cause some confusion between the two.
- Do we use the current sourcecode icon for the _Inline Code_ and then change the _Edit Source Code_ icon to the something else? For example, the `</>` icon or some form of the `{ }` icon? We feel the _Edit Source Code_ icon can be the more complex of the icons, as it a more robust feature of code editing and not simply an HTML tag.
- Do we find a different icon for the _Inline Code_ and leave the _Edit Source Code_ as is? If that's the case, what icon is best for the _Inline Code_? - this may be something to consider, if changing the current icon will cause familiarity issues for people already using the WYSIWYG for some time. Suggestions for the _Inline Code_ icon are then welcome, in that case; we can update the thread with some ideas, if this is what is required and if it's the appropriate place to do so (?)

Okay, I think that's all we have. 

Thank you!

**Comment by Natetronn** [comment]
I wanted to add other thoughts I had while working on this PR, this to be thorough and or for posterity:

You can see a list of [available TinyMCE toolbar buttons](https://www.tiny.cloud/docs/tinymce/latest/available-toolbar-buttons/).

The (inline) Code button (not to be confused with the Code plugin that is used for the Edit Source feature) is available in their API (and is how we accessed it), as well as being an item in the dropdown Menus. It isn't available as a toolbar button, however, as far as we know. That's why we're having to create it manually.

That said, I can't help but think this should be (and/or) a Feature Request or a Pull Request on TinyMCE proper at some point. I realize that isn't necessarily related to Directus itself, but I did want to mention it. Particularly depending on Directus' stance for features that could exist in Directus proper and or also in 3rd party tools it make use of.

Also, I'm going to leave the following here, just in case, as it was a relevant piece of the puzzle to get a working solution; it may be useful for others to better understand how to add toolbar buttons in the future:

https://github.com/tinymce/tinymce/issues/4898 - in particular, the Fiddle mentioned in the second comment: https://fiddle.tiny.cloud/qEgaab/3

**Comment by Natetronn** [comment]
I don't want to slow progress for @Abdallah-Awwad 's features request, but at the same time, I can't help but think we should also be discussing the HTML pre tag `<pre>`, as it is often used to wrap the HTML code tag `<code>`.

Like the code tag, it doesn't appear that there is a pre tag available as a toolbar button. There is, however, a pre tag available in the menus, so in theory, it could be added the same way we added code, if need be?

![image](https://github.com/user-attachments/assets/c5251b7a-5bc4-4df9-858a-5a330c80a3d1)

**Comment by rijkvanzanten** [comment]
> In regards to translations, do we add the translation string/value to the en-US.yaml only for the Interface? Is that all that is required? 

Yup! 

> Or do we add it to all the translations? 

This is a bit of a case by case gut feeling decision. If it's a very generic term, like "readonly" it makes more sense to keep it global. If it's something that'll only ever be used within the context of this interface, keeping it namespaced as such makes more sense

> If adding it to all, do we just leave its value in English, to later be translated on Crowdin or do we leave the value blank? 

Just add the English one and nothing else! Crowdin uses the en-US as the source read version, and syncs all other translations back as PRs in the `translations` branch.

> What's the best practice for submitting code updates to Directus proper as it relates to translations? The docs didn't mention this under [Contributions](https://docs.directus.io/contributing/introduction.html) or [Directus Local Setup](https://docs.directus.io/contributing/running-locally.html).

I believe this pertains to the questions above that I hopefully answered 🙂 CC @phazonoverload for making sure we add this into the docs later on 🙂

**Comment by rijkvanzanten** [comment]
> Do we use the same icon for both toolbar buttons? Our opinion was no, since that would cause some confusion between the two.

Ideally not, as it would indeed get confusing.

> Do we use the current sourcecode icon for the Inline Code and then change the Edit Source Code icon to the something else? For example, the </> icon or some form of the { } icon? We feel the Edit Source Code icon can be the more complex of the icons, as it a more robust feature of code editing and not simply an HTML tag. Do we find a different icon for the Inline Code and leave the Edit Source Code as is? If that's the case, what icon is best for the Inline Code? - this may be something to consider, if changing the current icon will cause familiarity issues for people already using the WYSIWYG for some time. Suggestions for the Inline Code icon are then welcome, in that case; we can update the thread with some ideas, if this is what is required and if it's the appropriate place to do so (?)

I'm a little worried that repurposing the existing icon might confuse current users 🤔 Maybe `codesample` is the next best thing? `{;}`

**Comment by rijkvanzanten** [comment]
> That said, I can't help but think this should be (and/or) a Feature Request or a Pull Request on TinyMCE proper at some point. I realize that isn't necessarily related to Directus itself, but I did want to mention it. Particularly depending on Directus' stance for features that could exist in Directus proper and or also in 3rd party tools it make use of.

Yeah it's a bit odd. They have some functionality only available as a sub-option in a dropdown of other toolbar options 🤔 I'm sure there's some design decision there that controls what's worth having a toolbar thing for vs just being an "exec" thing

**Comment by rijkvanzanten** [comment]
> Like the code tag, it doesn't appear that there is a pre tag available as a toolbar button. There is, however, a pre tag available in the menus, so in theory, it could be added the same way we added code, if need be?

Hmmmmmmmmmmmmmmmmm remind me.. `<code>` is inline; `<pre>` is block level? I'm wondering if it's the same button but it depends on whether you select an inline selection or a block selection 🤔

**Comment by Natetronn** [comment]
> I'm a little worried that repurposing the existing icon might confuse current users 🤔 Maybe `codesample` is the next best thing? `{;}`

We noticed that was available, but we both _very much didn't like it at all_ lol. We did consider it, since it's the only other code related icon available in TinyMCE proper. We can make it anything we want with any custom svg icon, though. We've been looking over some of the open source ones.

We can post up some other options, if you like?

**Comment by Natetronn** [comment]
> Hmmmmmmmmmmmmmmmmm remind me.. `<code>` is inline; `\<pre\>` is block level?

`<pre>` is indeed block-level: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/pre

> The `<pre>` HTML element represents preformatted text which is to be presented exactly as written in the HTML file. The text is typically rendered using a non-proportional, or monospaced, font. Whitespace inside this element is displayed as written.
> By default, `<pre>` is a block-level element, i.e. its default display value is block.

> I'm wondering if it's the same button but it depends on whether you select an inline selection or a block selection 🤔

I'm not sure? @Abdallah-Awwad would you mind researching this, please?

**Comment by Abdallah-Awwad** [comment]
> Hmmmmmmmmmmmmmmmmm remind me.. `<code>` is inline; `<pre>` is block level? I'm wondering if it's the same button but it depends on whether you select an inline selection or a block selection 🤔

> I'm not sure? @Abdallah-Awwad would you mind researching this, please?

> `<pre>` elements commonly contain [`<code>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/code), [`<samp>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/samp), and [`<kbd>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/kbd) elements, to represent computer code, computer output, and user input, respectively.

The `<pre>` tag functions similarly to the textarea element in terms of handling whitespace, making it primarily useful for representing block code. As far as I know, this is the main advantage we can derive from it.

Following the suggestion by @Natetronn, I tested the functionality on Directus WYSIWYG to verify its support for code blocks, and it works as expected (but need some enhancements). Additionally, I observed that the `<pre>` tag alone can serve as a code block without the need for the `<code>` tag.

If we aim to enhance WYSIWYG to support block code, we can consider submitting a new feature request. This enhancement would involve implementing the `<pre>` tag with potential enhancements to work as code block.

**Comment by Abdallah-Awwad** [comment]
> We can post up some other options, if you like?

Hello @rijkvanzanten once again, following up on what @Natetronn said, we've summarized the options with the following:

#### Option 1:

- Inline-code: `<>`
- Edit Source code: `</>`

#### Option 2:

- <div hspace="50"> Inline-code <img src="https://github.com/user-attachments/assets/c0365951-1a44-415b-b6c8-2832065e41a4" align="top" title="code_blocks"> </div>

- Edit Source code: `<>`


#### Option 3:

- Inline-code: `<>`
- Edit Source code: `{}` 

#### Option 4:

- Inline-code: `{}`
- Edit Source code: `<>`

#### Some context: 
- `<>`: Used as an inline code icon by GitHub, Slack, and Discord. TinyMCE also uses this icon for both inline code and edit source. Trello uses it for code snippets.
- `</>`: This icon is used by Teams for code snippets and by Notion for inline code.
- `{}`: This icon is used by Stack Overflow and Medium for code snippets.

Feel free to let us know if you need more options or additional context!

**Comment by rijkvanzanten** [comment]
I think options 1 and 2 are too similar and could easily be confusing. I think we'll have to go with either option 3 or 4. It's a complete personal opinion, but I prefer option 4. My reasoning is that the source of the WYSIWYG is HTML which uses `<>` at all times, whereas `{}` is a generic block scope in most other programming languages.

**Comment by Natetronn** [comment]
Just for sake of discussion, the pre tag is often used to wrap the code tag for blocks of code. I've seen `pre` and `<pre>` "icons" before. Maybe a bit too long, but perhaps just `code` or `<code>` could work, particularly if `pre` or `<pre>` is too be added at some point as well.

Another option, which may go against the current "style" of icons only (assumption), is to have a drop down for Pre, Code and Code Block (Pre/Code).

Which brings up the topic of Pre, Code and Code Blocks. I can't help but think Code shouldn't be added until Pre is added as well. Why? I'll show you...

This is the code tag on its own, as it's meant to be used; inline:

_(Please disregard the icons for now)_

![image](https://github.com/user-attachments/assets/f97dbd46-3d0a-4136-b124-3408dc0e406a)

And this is how the code tag is use on its own, but incorrectly; it stands to reason that this will happen by users at some point and cause issue:

![image](https://github.com/user-attachments/assets/ace17380-54b0-4488-9f2d-96e7767cde9f)

To address that issue, one would use a Code Block instead i.e. Pre wraps Code:
![image](https://github.com/user-attachments/assets/877a8d2c-e1bb-4e67-a821-7e9715d74ee0)

Note: according to Abdullah, the code tag is being stripped out of the pre tag upon save. I'm hoping this is an easy fix and just some sort of "allowed tags feature" within TinyMCE and we just flip that switch, so to speak, versus some more complicated bug. We've yet to investigate that, though. And I need to confirm his finding.

Point is, us adding Code (inline) creates the potential for it's misuse and adds some potential issues when in the hands of users and editors who don't know the difference between Inline Code and Code Blocks.

That said, we're wondering about your thoughts on this and if we should proceed with the code alone or should we reevaluate the current goal to further include Pre and Code Block? Or should we stop before going too far down this "rabbit hole", since it's starting to be a bit more of a "can of worms" than initially anticipated?

**Comment by rijkvanzanten** [comment]
Oh my.. I was hoping that TinyMCE would automagically use a `<pre>` when your selection would be multiline, but it looks like it basically creates multiple paragraphs with `<code>` in it? 😐 

> That said, we're wondering about your thoughts on this and if we should proceed with the code alone or should we reevaluate the current goal to further include Pre and Code Block? Or should we stop before going too far down this "rabbit hole", since it's starting to be a bit more of a "can of worms" than initially anticipated?

Haha oh the typical joys of implementing something in the Directus project. I'm similarly conflicted. It feels like a logical move to support "inline/block code snippet support" as a single piece of work, but it did indeed scope creep quite a bit from the original "Support `<code>`".

I think your correct assessment of the potential misuse / issues around code and pre is enough of a deciding factor that I think we should support both.

That being said.. I wonder if there's ever a case where somebody intentionally wants to use `<pre>` for a selected inline word, or wants to use multiple `<code>` for a multi-line section of code. Maybe there's a way where we can support both those tags under the same button? That way we'll insert a `<code>` if you select a section of a line of text, and use `<pre>` if you selected multiple lines? 🤔

**Comment by Natetronn** [comment]
> That being said.. I wonder if there's ever a case where somebody intentionally wants to use `\<pre\>` for a selected inline word, or wants to use multiple `<code>` for a multi-line section of code.

I don't think they'll intentionally want do that. But I do think they'll just paste in the code and then make it a `<pre>`  or a `<code>` or `<pre><code>` after the fact, which may cause some issues thereafter.

> Maybe there's a way where we can support both those tags under the same button? That way we'll insert a `<code>` if you select a section of a line of text, and use `\<pre\>` if you selected multiple lines? 🤔

I thought about that (we discussed it even) and were wondering if that should be how it works. If one line is selected, the code icon (whatever it may be) sets `<code>` and if multiple lines are selected, that same button sets it as `<pre><code>`.

That's how the `<>` icon works here on Github and wouldn't be a bad solution; albeit, that's a bit harder to implement and breaks out of the TinyMCE paradigm (api) and into custom functionality (I'd assume; I wonder if TinyMCE knows when a single line vs multi-line is selected and has that feature expose for us to make use of.)

Just to be thorough, using Github, Markdown and Discord as examples, three open/close backticks is the equivalent of pre wraps code aka Code Block. One open/close back tick is the same as code (inline).

Probably not as often, but I read Pre can be used standalone for poems, for example. So it isn't always combined with code based text and the code tag. Kind of why I was thinking Pre, Code and Pre/Code (Code Blocks) should all be available. But of the three, Pre standalone could come later or perhaps not at all; just stick to inline code and code blocks.

The issue with `<code>` is some users might read that and not realize it's really `"<inline-code>"` (or even understand what inline code is to begin with). Having the additional Code Block feature can help them understand the difference exists, since they'd have two options to choose from. That is, I think people will then realize there is a difference and maybe get the hint. Maybe it could be Code (Inline) and Code (Block) or something to make it more clear. 

Then again, I guess that wouldn't matter anymore if we combined the feature for code (inline) and code (block) into one "automated" feature like Github did.

**Comment by rijkvanzanten** [comment]
I'm finding myself come back to the issue that I don't think that the general audience really understands the difference between the three 🤔 

I think the GitHub approach—while a bit more complicated to implement—is the right move from a user experience design perspective 🙂

**Comment by Natetronn** [comment]
> I'm finding myself come back to the issue that I don't think that the general audience really understands the difference between the three 🤔
> 

For sure. I agree that they most likely do not understand. Some developers may have forget themselves, too. 

But just to play devil's advocate here, however, we can place that decision on the developers that do know. They can turn on any of the three options (or not) on a case by case basis. So it's giving them freedom to choose if it's appropriate to use and when, depending on their context. That said, I do agree with your next point, so moving on to that.

> I think the GitHub approach—while a bit more complicated to implement—is the right move from a user experience design perspective 🙂

For sure. It works very well here on Github and "kills two birds with one stone" too. 

I think that Github has the right icon for the job, too. Which brings us a bit full circle to the fact that the Edit Source Code icon _should_ be changed, in my humble opinion. I realize that may cause some grief for people who don't like change, but it is what it is and right is right (subjective/opinion, but hear me out). 

That same icon is used for the Block Editor (although all those icons are wrapped in a square):

![image](https://github.com/user-attachments/assets/4aec606f-2d99-40e4-aea4-86d5fb5ed0cc)

And that same icon is also used in the Markdown Editor. And by the way, the Markdown Editor uses the same single button implementation for inline code and code blocks that Github does:

![image](https://github.com/user-attachments/assets/7426b1e2-80c7-49fb-9643-69eef43350c9)

So the Edit Source Code icon is really the "odd man out" and should be the one that changes, not the other way around.

Anyway, I agree that the single button for code and code block makes the most sense. 

Aside: We can also add in standalone Pre at a later time in a different PR, if we find that people want it. The icon for that could just be the word `pre` or maybe `<pre>`. That's an easy one, but less important.

**Comment by rijkvanzanten** [comment]
> So the Edit Source Code icon is really the "odd man out" and should be the one that changes, not the other way around.

Deal 🤝

**Comment by Abdallah-Awwad** [comment]
Hey @rijkvanzanten 
Apologies for the long delay on this PR. Unexpected priorities came up, but I’ve now completed it and tried to ensure it meets the expected standards. Let me know if anything needs adjustment, I’m ready for any feedback. 
Thanks, and sorry once again!

**Comment by Maladroy** [comment]
Hi, any status on this feature? My content team would really benefit from this tiny addition, as they are not familiar with HTML (and they refused to use markdown).

**Reply by Abdallah-Awwad** [reply]
Hi @Maladroy, super sorry for the late reply! 
Just wanted to make sure you are aware it's live since 11.10.0
